### ISSUE TITLE
Overload inference with `*args`

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#8645]
I first asked this here: https://github.com/python/typing/discussions/1828, but I am starting to think this might be a bug in pyright (edge case), because when one removes the `def foo() -> tuple[()]` overload then the variable size overload gets selected, according to the [documented resolution](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=overloads)

> Exception 1: When an *args (unpacked) argument matches a *args parameter in one of the overload signatures, this overrides the normal order-based rule.

Code sample in [pyright playground](https://pyright-play.net/?strict=true&reportMissingModuleSource=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCGAzkQKZ4D68CpANFGAG7kA2YBAJgFBcACTrdhyhQAxFBQBeAAxcOpYFGBgwACgCUUALQA%2BWAFcELUgG0NAXQBcUAHR2%2BAkG04jxUgIxyFSlaoAe1qgw9AD0mroGRqZBVrb2-MxOQq4SkgBMXorKan7ugSjBUH5p%2BYVh2nowhsYmQfQx1nY2DonOwmISUIwEIEgEAEbGmT5qAFQBmAXhldXRBfRNsU3D2arjpdORNXVxNlZcIiIgpFUgKEU8xGSU1KSqq%2Br0VVFm6uaaHYCg5FxX5DBUiDuq3cjy2cxg7xS31%2BN0B9187noaVBz2280mEI%2B4mhJD%2BAJo8LUiLStAAzCjZrV0YssVAcdd-rdCWsTMSye8npSdjSUoAZcigxwAxqQkMxhAAiVGmCzirhAA)

```python
from typing import assert_type, overload

@overload  # n=0
def foo() -> tuple[()]: ...
@overload  # n=1
def foo(x: int, /) -> tuple[int]: ...
@overload  # n=2
def foo(x1: int, x2: int, /) -> tuple[int, int]: ...
@overload  # n variable
def foo(*x: int) -> tuple[int, ...]: ...
def foo(*x: int) -> tuple[int, ...]:
    return x

assert_type(foo(), tuple[()])  # ✅
assert_type(foo(1), tuple[int])  # ✅
assert_type(foo(1, 2), tuple[int, int])  # ✅
assert_type(foo(1,2,3), tuple[int, ...])  # ✅
assert_type(foo(*[1,2,3]), tuple[int, ...])  # ❌ received "tuple[()]"

```

<details><summary>Example without the empty signature overload</summary>

Code sample in [pyright playground](https://pyright-play.net/?strict=true&reportMissingModuleSource=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCGAzkQKZ4D68CpANFGAG7kA2YBAJgFBcDEUAASat2HKFH4oAvAAY%2BUDqWBRgYMAAoAlFAC0APlgBXBC1IBtLQF0AXFAB0DrkOYg2ncZKkBGLouWqNAA9bVBh6AHptfSMTc1Cbe0dnEXcJKGkAJl8lFTV1QK8QlDCoQIyiksjdAxhjUzNQ%2BnjbBzsnYVdRD3SoRgIQJAIAI1Ns-zyAKmDMYqiaurji%2BlaE1rHcjSmKuZj6xsS7Gy5xcRBSWpAUUp5%2BYjJKalJ1AK16WtiLTUttNMBQci47uQYFREE8Xl5NG8Fg1it9uv9AQ9Qc88l56BlIbtFiV4j9%2BAiSECQTQURo0RlaABmTHvPZLA5wv4AwlIkkvCZmclU75Qj77FZ4qD-IA)

```python
from typing import assert_type, overload

# @overload  # n=0
# def foo() -> tuple[()]: ...
@overload  # n=1
def foo(x: int, /) -> tuple[int]: ...
@overload  # n=2
def foo(x1: int, x2: int, /) -> tuple[int, int]: ...
@overload  # n variable
def foo(*x: int) -> tuple[int, ...]: ...
def foo(*x: int) -> tuple[int, ...]:
    return x

# assert_type(foo(), tuple[()])  # ✅
assert_type(foo(1), tuple[int])  # ✅
assert_type(foo(1, 2), tuple[int, int])  # ✅
assert_type(foo(1,2,3), tuple[int, ...])  # ✅
assert_type(foo(*[1,2,3]), tuple[int, ...])  # ✅
```

The `*args` overload gets correctly selected.

</details>

**Comment by erictraut** [comment]
As you're probably aware, the behaviors for overload matching is not currently specified in the typing spec. In my opinion, this is one of the highest-priority areas for us [to specify](https://docs.google.com/presentation/d/1ash1ellodiV6FTyVnY-Q6V2G5vJ68ycWBr6MYkayC1U/edit#slide=id.g2770f948faa_0_5). However, I recognize that specifying the behavior here is going to be difficult.

Until we have a specification for overload resolution, I don't plan to modify pyright's current behavior unless there is a clear bug. The behavior you're seeing here is not a clear bug; it's a consequence of the algorithm pyright uses.

For an explanation of pyright's current algorithm, see [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=overloads).

For comparison, here's [mypy's documentation for overloads](https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-calls-to-overloads). According to mypy's documentation, it should produce the same results as pyright's algorithm. In particular, it says "If there are multiple equally good matching variants, mypy will select the variant that was defined first.". It lists two exceptions to this rule, neither of which apply in this case. So mypy's documentation is apparently not complete or there is a bug where the implementation doesn't match the documented behavior.

**Comment by randolf-scholz** [comment]
Eric, thank you for the detailed response. I understand that overload specification is really a pain point. However, I disagree with your analysis. The *"pick first overload rule"* is not directly applicable here. The `mypy` documentation you quoted clearly states, in the very first sentence:

> When you call an overloaded function, mypy will infer the correct return type by picking the best matching variant, **after taking into consideration both the argument types and arity.**

In my example, there are overloads for `foo` with nullary, unary, binary and variadic signature (all positional-only). So according to this rule, it makes sense that `foo()` should select the nullary, `foo(1)` the unary, `foo(1,2)` the binary and `foo(1,2,3)` and `foo(*[1,2,3])` the variadic overload variant. Only if there were multiple variants for some arity, the `pick first match` rule should come into play.

-----

So why does `pyright` select the nullary variant, given a variadic argument?
When I try following pyright's overload specification step by step, I arrive at the same conclusion (that this is a bug):

> 1. Pyright first filters the list of overloads based on simple “arity” (number of arguments) and keyword argument matching. For example, if one overload requires two positional arguments but only one positional argument is supplied by the caller, that overload is eliminated from consideration. Likewise, if the call includes a keyword argument but no corresponding parameter is included in the overload, it is eliminated from consideration.

So for the argument `*list[int]`, which is unknown size, this cannot eliminate any of the overloads in my example.

> 2. Pyright next considers the types of the arguments and compares them to the declared types of the corresponding parameters. If the types do not match for a given overload, that overload is eliminated from consideration. Bidirectional type inference is used to determine the types of the argument expressions.

Here, the argument type is `*list[int]` ≅ `*tuple[int, ...]`, which matches all the overload variants, which can equivalently be expressed as `*tuple[()]`, `*tuple[int]`, `*tuple[int, int]` and `*tupe[int, ...]`.

> 3. If only one overload remains, it is the “winner”.

All overloads should still remain at this point.

> 4. If more than one overload remains, the “winner” is chosen based on the order in which the overloads are declared. In general, the first remaining overload is the “winner”. There are two exceptions to this rule. Exception 1: When an *args (unpacked) argument matches a *args parameter in one of the overload signatures, this overrides the normal order-based rule. Exception 2: When two or more overloads match because an argument evaluates to Any or Unknown [...]

So by exception 1, the overload `def foo(*x: int) -> tuple[int, ...]` should be declared the winner.

What am I missing? And why would it not apply when there is no nullary overload?

Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCGAzkQKZ4D68CpANFGAG7kA2YBAJgFBcACTrdhyhQAxFBQBeAAxcOpYFGBgwACgCUUALQA%2BWAFcELUgG0NAXQBcUAHR2%2BAkG04jxUgIxyFSlaoAe1qgw9AD0mroGRqZBVrb2-MxOQq4SkgBMXorKan7ugSjBUH5p%2BYVh2nowhsYmQfQx1nY2DonOwmISUIwEIEgEAEbGmT5qAFQBmAXhldXRBfRNsU3D2arjpdORNXVxNlZcIiIgpFUgKEU8xGSU1KSqq%2Br0VVFm6uaaHRCIcNaAoOT0CDgvTQAAsYH8uFdyDAqIg7qt3I8tnMYO8Ul9AX8AUCkKDwVBfpCSNDYTR7r53PQ0kjntt5pNUR9xBifgTscCwRCoTc4eS1JSoGl6ABmGmzWr0xZMqAsrFQQEc-GE7kw258tYmAVCqDC95PcU7KXo75yhW4zlQQAy5DxxAB3JAwEFgfT4FD6FgsHpwBitISWFqCFwdDzDfo9fylUKbWkopbxRxtFJSDLyRRhkD%2BPIM%2BjFSNQcoRGMSwoNXYBpJBtxdHp9QakUPh9YM6MGyV2OPNVNQdNrCZBFsvQ3t-2HKDHU7nPyXYk8sk9xH6wcFNGfE1s%2BU4vFcmeq3nzqlipcl5fS2Xrs1bglE667ufhrUiw90wpG1eY8%2Bbi3Kneku490aalSIp6sixYLO2p5rv8G6KhCQA), [mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=4391750d5372d4126100c9777aa7c2fc)

```python
from typing import assert_type, overload

@overload  # n=0
def foo() -> tuple[()]: ...
@overload  # n=1
def foo(x: int, /) -> tuple[int]: ...
@overload  # n=2
def foo(x1: int, x2: int, /) -> tuple[int, int]: ...
@overload  # n variable
def foo(*x: int) -> tuple[int, ...]: ...
def foo(*x: int) -> tuple[int, ...]:
    return x

assert_type(foo(), tuple[()])  # mypy: ✅, pyright: ✅
assert_type(foo(1), tuple[int])  # mypy: ✅, pyright: ✅
assert_type(foo(1, 2), tuple[int, int])  # mypy: ✅, pyright: ✅
assert_type(foo(1, 2, 3), tuple[int, ...])  # mypy: ✅, pyright: ✅
assert_type(foo(*[1, 2, 3]), tuple[int, ...])  # mypy: ✅, pyright: ❌

# without nullary overload:
@overload  # n=1
def bar(x: int, /) -> tuple[int]: ...
@overload  # n=2
def bar(x1: int, x2: int, /) -> tuple[int, int]: ...
@overload  # n variable
def bar(*x: int) -> tuple[int, ...]: ...
def bar(*x: int) -> tuple[int, ...]:
    return x

assert_type(bar(1), tuple[int])  # mypy: ✅, pyright: ✅
assert_type(bar(1, 2), tuple[int, int])  # mypy: ✅, pyright: ✅
assert_type(bar(1, 2, 3), tuple[int, ...])  # mypy: ✅, pyright: ✅
assert_type(bar(*[1, 2, 3]), tuple[int, ...])  # mypy: ✅, pyright: ✅
```

**Comment by erictraut** [comment]
I agree there was a hole in my analysis, and there is indeed a bug in pyright's handling of zero-arity signatures in this case. If you delete (or temporarily comment out) the zero-arity case, you'll see that the type evaluation changes for `foo(*[1, 2, 3])`.

This bug will be addressed in the next release.

Thanks for persisting in helping me see the flaw in my analysis.

**Comment by erictraut** [comment]
This is included in pyright 1.1.375.
